apache/cloudstack
# Support network-scope settings and add setting to override vr service offering

### Description

As of now, it is possible to override the virtual router service offering with global, domain, and account level configurations. What if a customer wants to change the service offering for only one network? There isn't such an option in Cloudstack. As an admin, you can only override the network offering for a domain, account, or global, affecting all the networks in that domain, account, or zone. This pull request aims to add a network-level setting to override the virtual router service offering per network.















### Types of changes


### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale


### How Has This Been Tested?




To test this feature, please follow the below steps:

1. Create a network
2. Create VM in that network (to create a VR)
3. List the network and get the UUID of the network
4. Check the service offering of the network
5. Create a new router service offering
6. Set the `network.router.service.offering` to the UUID of the service offering for the network using the cmk
7. Restart the network with clean up
8. Check the service offering of the network

Useful `cmk` commands:

```
(localcloud) 🐱 > list networks filter=id,name
{
  "count": 2,
  "network": [
    {
      "id": "c9d70872-b877-4f02-92b5-b4b08e1095d7",
      "name": "isolated-network"
    },
    {
      "id": "1ebf1ca9-7930-4f2f-8263-8da1840f02fc",
      "name": "shared"
    }
  ]
}
(localcloud) 🐱 > list configurations networkid=c9d70872-b877-4f02-92b5-b4b08e1095d7
{
  "configuration": [
    {
      "category": "Advanced",
      "description": "Uuid of the service offering used by virtual routers of this network; if NULL - account/system offering will be used",
      "isdynamic": true,
      "name": "network.router.service.offering",
      "scope": "network",
      "value": ""
    }
  ],
  "count": 1
}
(localcloud) 🐱 > list routers filter=serviceofferingid,serviceofferingname
{
  "count": 1,
  "router": [
    {
      "serviceofferingid": "f37ed5fd-6f26-4316-8b27-dbebb0b56a5a"
      "serviceofferingname": "System Offering For Software Router"
    }
  ]
}
(localcloud) 🐱 > update configuration networkid=c9d70872-b877-4f02-92b5-b4b08e1095d7 name="network.router.service.offering" value="5ad7f2c4-29e0-460b-901b-05ee8de8a0df"
{
  "configuration": {
    "category": "Advanced",
    "description": "Uuid of the service offering used by virtual routers of this network; if NULL - account/system offering will be used",
    "isdynamic": true,
    "name": "network.router.service.offering",
    "scope": "network",
    "value": "5ad7f2c4-29e0-460b-901b-05ee8de8a0df"
  }
}
(localcloud) 🐱 > restart network id=c9d70872-b877-4f02-92b5-b4b08e1095d7 cleanup=true
{
  "success": true
}
(localcloud) 🐱 > list routers filter=serviceofferingid,serviceofferingname
{
  "count": 1,
  "router": [
    {
      "serviceofferingid": "5ad7f2c4-29e0-460b-901b-05ee8de8a0df"
      "serviceofferingname": "Medium Instance"
    }
  ]
}
```
